text_extensions_for_pandas/io/watson/nlu.py

```diff
@@ -365,54 +346,67 @@ def _make_relations_dataframe_zero_copy(relations):
-    parent_cols = [[] for i in range(num_parent_cols)]
-    begins = []
-    ends = []
-    confidences = []
-    for e in entities:
-        for m in e["mentions"]:
-            for i in range(num_parent_cols):
-                parent_elem = e[_entity_mentions_parent_names[i]]
-                parent_cols[i].append(parent_elem)
-            begins.append(m["location"][0])
-            ends.append(m["location"][1])
-            confidences.append(m["confidence"])  # N.B. confidence of mention, not entity
-    # Construct columns, then convert to a DataFrame
-    df_cols = {
-        _entity_mentions_parent_names[i]: parent_cols[i]
-        for i in range(len(_entity_mentions_parent_names))
-    }
-    df_cols["span"] = SpanArray(original_text, begins, ends)
-    df_cols["confidence"] = confidences
-    return util.apply_schema(pd.DataFrame(df_cols), _entity_mentions_schema,
-                             apply_standard_schema)
+    if len(entities) == 0:
+        return pd.DataFrame(), pd.DataFrame()
+
+    table = util.make_table(entities)
+
+    # Check if response includes entity mentions
+    mention_name_cols = [(name, table.column(name)) for name in table.column_names
+                         if name.lower().startswith("mentions")]
+
+    # Make entities and entity mentions (optional) DataFrames
+    if len(mention_name_cols) > 0:
+        mention_names, mention_cols = zip(*mention_name_cols)
+
+        # Create the entities DataFrame with mention arrays dropped
+        table = table.drop(mention_names)
+        pdf = table.to_pandas()
+
+        # Flatten the mention arrays to be put in separate table
+        mention_arrays = [pa.concat_arrays(col.iterchunks()) for col in mention_cols]
+        flat_mention_arrays = [a.flatten() for a in mention_arrays]
+        table_mentions = pa.Table.from_arrays(flat_mention_arrays, names=mention_names)
+
+        # Convert location/text columns to span
+        location_col, location_name = util.find_column(table_mentions, "location")
+        text_col, text_name = util.find_column(table_mentions, "text")
+        if original_text is None:
+            raise ValueError(
+                "Unable to construct target text for converting entity mentions to spans")
+
+        char_span = util.make_char_span(location_col, text_col, original_text)
+        table_mentions = table_mentions.drop([location_name, text_name])
+
+        # Create the entity_mentions DataFrame
+        pdf_mentions = table_mentions.to_pandas()
+        pdf_mentions["span"] = char_span
+
+        # Align index of parent entities DataFrame with flattened DataFrame and ffill values
+        mention_offsets = mention_arrays[0].offsets.to_numpy()
+        pdf_parent = pdf.set_index(mention_offsets[:-1])
+        pdf_parent = pdf_parent.reindex(pdf_mentions.index, method="ffill")
+
+        # Add columns from entities parent DataFrame
+        pdf_mentions["type"] = pdf_parent["type"]
```

Comment:
I seems like we only want "type" column from the entities df, unless "relevance" is useful too?

Reply:
The set of columns from the "entities" result that we include in the "entity_mentions" DataFrame is something of a judgement call. On one hand, including additional fields would save the user the trouble of calling `pd.merge()` on the two DataFrames. On the other hand, the properties in the parent entity record are associated with document-level entities, and we don't want to create the impression of a direct connection to individual mentions. There's also the issue of cluttering the entity mentions DataFrame with extra columns.

Here's an example. If you run this code:
```python
natural_language_understanding.analyze(
    text="I hate John Smith. You love John Smith.",,
    return_analyzed_text=True,
    features=nlu.Features(
        entities=nlu.EntitiesOptions(sentiment=True, mentions=True, emotion=True)
    )).get_result()
```
you get back this result:
```
{'usage': {'text_units': 1, 'text_characters': 39, 'features': 1},
 'language': 'en',
 'entities': [{'type': 'Person',
   'text': 'John Smith',
   'sentiment': {'score': -0.296829, 'mixed': '1', 'label': 'negative'},
   'relevance': 0.9999,
   'mentions': [{'text': 'John Smith',
     'location': [7, 17],
     'confidence': 0.995878},
    {'text': 'John Smith', 'location': [28, 38], 'confidence': 0.992923}],
   'emotion': {'sadness': 0.112204,
    'joy': 0.289177,
    'fear': 0.061407,
    'disgust': 0.134739,
    'anger': 0.409954},
   'count': 2,
   'confidence': 0.999971}],
 'analyzed_text': 'I hate John Smith. You love John Smith.'}
```

Note how there are two mentions of "John Smith" with very different local sentiment, but NLU associates both of these with the same document-level sentiment information (`sentiment': {'score': -0.296829, 'mixed': '1', 'label': 'negative'}`) and document-level emotion information (`'emotion': {'sadness': 0.112204,...`). Also, the response includes **three** confidence values: One for each of the entity mentions and one for the document-level entity "John Smith".

Reply:
Yeah, sounds good. Thanks for the great explanation! I'll include this in the tests.